client/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+    FailedToParseTransactionPayload,
+    /// When the given proof is invalid.
+    #[error("invalid proof given: got merkle root of {0} but expected {1}")]
+    InvalidProof(String, String),
+    /// When the given message is well decoded and verified, but the message argument is invalid.
+    #[error("invalid message argument given: {0}")]
+    InvalidMessageArgument(String),
+    /// When the relayer account has not enough balance to execute the transaction.
+    #[error("not enough balance: got {0}")]
+    NotEnoughBalance(u64),
+    /// When the account sequence given in the transaction is invalid.
+    #[error("invalid account sequence; expected {0} but got {1}")]
+    InvalidAccountSequence(u64, u64),
+    /// When the contract sequence given in the transaction is invalid.
+    #[error("invalid contract sequence; expected {0} but got {1}")]
+    InvalidContractSequence(u64, u64),
+    /// When the contract fails to execute the transaction with its own error.
+    #[error("internal contract error: {0}")]
+    InternalContractError(String),
+    /// When the contract is missing.
+    #[error("no such contract: {0}")]
+    NoSuchContract(String),
+    /// An unknown error.
+    #[error("unknown error: {0}")]
+    Unknown(String),
+}
+
+/// An abstract information about a block.
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone)]
+pub struct Block {
+    /// The height of the block.
+    pub height: u64,
+    /// The UNIX timestamp of the block in seconds.
+    pub timestamp: u64,
+}
+
+/// An abstraction of the residential chain with its treasury deployed on it.
+///
+/// One trivial implementation of this trait would carry the API endpoint of the full node and
+/// the relayer account used to submit message delivering transactions.
+#[async_trait::async_trait]
+pub trait ResidentialChain: Send + Sync {
+    /// Returns the name of the chain.
+    async fn get_chain_name(&self) -> String;
+
+    /// Checks whether the chain is healthy and the full node is running.
+    async fn check_connection(&self) -> Result<(), Error>;
+
+    /// Getes the latest finalized block on the chain.
+    async fn get_last_block(&self) -> Result<Block, Error>;
+
+    /// Returns the address and the current balance (which is used to pay the gas fee) of the relayer account in this chain.
+    ///
+    /// Note that there is no authority of the relayer account; it is just a convenient account to pay the gas fee
+    /// (i.e., there is no need to check the transaction submitter by the contract).
+    async fn get_relayer_account_info(&self) -> Result<(String, Decimal), Error>;
+
+    /// Returns the latest header that the light client has verified.
+    async fn get_light_client_header(&self) -> Result<Header, Error>;
+
+    /// Returns the current balance of fungible tokens in the treasury contract.
+    async fn get_treasury_fungible_token_balance(&self) -> Result<HashMap<String, Decimal>, Error>;
```

Comment:
I think it returns all token balances in the treasury.
Do we need to have a token array in solidity for this? (e.g. arr[0] = usdt_address, arr[1] = usdc_address)

It would be better to call it with the token_address argument and return <String, Decimal> for one token.

Reply:
fixed